### PR TITLE
random sampling of dij for speedup of dose calculation of the optimization

### PR DESCRIPTION
**Pull request by amitantony**
issue with the previous idea of dose grid speed up was that our carbon ion spot could be smaller than the voxel. 
here we sample some percentage of voxels from each structure, but maintain a ring of voxels on the inside of each stucture for full resolution of edge effects, 

Here selecting samples from a full dij ( to be kept for use in final dose calculation ) 

ideas that can be further developed:
1. I can think of scalar mulitplication with mask vector for voxel selection.
2. also can define unique sampling fractions for each structure if needed ( eg. target (90%), OAR(50%), body( 10% ) )

### CODE REVIEW

**Comment by amitantony** [comment]
the tesingspeedup.m is a script to test the speed up

**Comment by becker89** [comment]
@amitantony can you have a look on why travis is failing ?

**Comment by wahln** [comment]
It's probably the missing octave graphics toolkit, that now needs to be installed explicitly. @ahmadnish fixed it at least on dev and master...

File: matRad_dijSampler.m [hunk 1]
```diff
@@ -0,0 +1,63 @@
+function [cst,dij] = matRad_dijSampler(cst,ct,dij, margin_size,sparCity)
+% marginsize given in pixel thickness
+% fraction of voxels to be sampled [ Target OAR]
+
```
Comment (markbangert) [inline matRad_dijSampler.m]: complete function header is missing. also I think this script could be moved to the tools folder

File: matRad_dijSampler.m [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+m = (2*margin_size)+1;
+msk = ones(m,m,m);    % mask for convolution
+
+volThresh = 1000;  % setting full sampling for small structures 
+idx = [];
+fullmap =[];            % just for displaying 
+s = RandStream('mlfg6331_64');                               % initialize random number generator (seed value )
+cst = matRad_setOverlapPriorities(cst,ct.cubeDim);
+
+for i= 1:size(cst,1)
```
Comment (markbangert) [inline matRad_dijSampler.m]: whitespace is missing..

File: matRad_dijSampler.m [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+volThresh = 1000;  % setting full sampling for small structures 
+idx = [];
+fullmap =[];            % just for displaying 
+s = RandStream('mlfg6331_64');                               % initialize random number generator (seed value )
+cst = matRad_setOverlapPriorities(cst,ct.cubeDim);
+
+for i= 1:size(cst,1) 
+    if  ~isempty(cst{i,6})
+        for j=1:numel(cst{i,6})
```
Comment (markbangert) [inline matRad_dijSampler.m]: again whitespaces are missing ;)

j = 1:numel...

File: matRad_dijSampler.m [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            cst{i,6}(j).numOfVoxels = numel(cst{i,4}{1});
+        end 
+    end
+end
+
+for i=1:size(cst,1)
```
Comment (markbangert) [inline matRad_dijSampler.m]: again...

File: matRad_dijSampler.m [hunk 5]
```diff
@@ -0,0 +1,63 @@
+function [cst,dij] = matRad_dijSampler(cst,ct,dij, margin_size,sparCity)
+% marginsize given in pixel thickness
+% fraction of voxels to be sampled [ Target OAR]
+
+m = (2*margin_size)+1;
+msk = ones(m,m,m);    % mask for convolution
+
+volThresh = 1000;  % setting full sampling for small structures
```
Comment (markbangert) [inline matRad_dijSampler.m]: better comment: % stuctures where the number of voxels is smaller that volThres will be ommitted from sampling.

File: matRad_dijSampler.m [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    end
+end
+
+for i=1:size(cst,1)
+    
+    % fraction of voxels sampled 
+    if strcmp(cst{i,3}, 'TARGET') || (numel(cst{i,4}{1}) < volThresh)                % threshold volume for small structures
+        sp = sparCity(1);
+    elseif strcmp(cst{i,3}, 'OAR')
+        sp = sparCity(2);                                   % sampling fraction
```
Comment (markbangert) [inline matRad_dijSampler.m]: if sparCity is the sampling fraction then why not name it samplingFraction?
and sp could be voiSamplingFraction...?

:)

File: matRad_dijSampler.m [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    
+    % fraction of voxels sampled 
+    if strcmp(cst{i,3}, 'TARGET') || (numel(cst{i,4}{1}) < volThresh)                % threshold volume for small structures
+        sp = sparCity(1);
+    elseif strcmp(cst{i,3}, 'OAR')
+        sp = sparCity(2);                                   % sampling fraction 
+    end
+    % getting the margins 
+    mVOI = zeros(ct.cubeDim) ;
+    mVOI( cst{i,4}{1}) = 1;
```
Comment (markbangert) [inline matRad_dijSampler.m]: whitespace too much

File: matRad_dijSampler.m [hunk 8]
```diff
@@ -0,0 +1,63 @@
+function [cst,dij] = matRad_dijSampler(cst,ct,dij, margin_size,sparCity)
```
Comment (markbangert) [inline matRad_dijSampler.m]: usually we use letter case-separated words e.g. marginSize for variable names in matRad...

File: matRad_dijSampler.m [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    if strcmp(cst{i,3}, 'TARGET') || (numel(cst{i,4}{1}) < volThresh)                % threshold volume for small structures
+        sp = sparCity(1);
+    elseif strcmp(cst{i,3}, 'OAR')
+        sp = sparCity(2);                                   % sampling fraction 
+    end
+    % getting the margins 
+    mVOI = zeros(ct.cubeDim) ;
+    mVOI( cst{i,4}{1}) = 1;
+    mVc = convn(mVOI,msk,'same');
+    mVd = mVc - mVOI.*sum(msk(:));   % finding closed edge voxels
```
Comment (markbangert) [inline matRad_dijSampler.m]: The approach with the convolution is not really conserving the geometry. Assume you have a VOI that is very cigar shaped, in the most extrem case just a single line of voxels extending along a single CT dimension. Then this convolution would basically wipe out the entire structure. It does not necessarily preserve the edges - or am I wrong?

File: matRad_dijSampler.m [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+ vec (fullmap)  = 1;
+selectionDiag = spdiags(vec, 0 , dij.ctGrid.numOfVoxels, dij.ctGrid.numOfVoxels ) ;
+
+tic,  dij.physicalDose{1} = ( dij.physicalDose{1}' * selectionDiag)';toc
+
+     if isfield(dij,'mAlphaDose')        
+        dij.mAlphaDose{1} = (dij.mAlphaDose{1}' * selectionDiag)';
+        dij.mSqrtBetaDose{1} = (dij.mSqrtBetaDose{1}' * selectionDiag)';        
+    end
+xx = zeros(ct.cubeDim);
```
Comment (markbangert) [inline matRad_dijSampler.m]: please look for better variable name

File: matRad_dijSampler.m [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    idx2 = datasample(s, list, floor(sp * numel(list)),'Replace',false );
+    cst{i,4}{2} = cst{i,4}{1};                                                      % saving the original idx ... required for ray casting !!!
+    cst{i,4}{1} = [idx; idx2];
+   
+    idxRm =  setdiff(cst{i,4}{2},cst{i,4}{1});
+    
+    fullmap = [ fullmap; idx; idx2];                                    % for plotting the map of selected voxels
+    
+end
+ vec = zeros( numel(mVOI),1);
```
Comment (markbangert) [inline matRad_dijSampler.m]: fix indenting.

File: matRad_dijSampler.m [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    % getting the margins 
+    mVOI = zeros(ct.cubeDim) ;
+    mVOI( cst{i,4}{1}) = 1;
+    mVc = convn(mVOI,msk,'same');
+    mVd = mVc - mVOI.*sum(msk(:));   % finding closed edge voxels
+    idx = find(mVd<0);   % idx
+    % randomly sampling within the margins
+    mVOI(idx) = 0;
+    list = find(mVOI);
+    idx2 = datasample(s, list, floor(sp * numel(list)),'Replace',false );
```
Comment (markbangert) [inline matRad_dijSampler.m]: https://de.mathworks.com/help/stats/datasample.html is part of the machine learning toolbox, is it? Please do not include additional dependencies besides image processing toolbox

**Pull request closed without merging**